luigi/contrib/hdfs/hadoopcli_clients.py

```diff
@@ -0,0 +1,288 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""
+The implementations of the hdfs clients. The hadoop cli client and the
+snakebite client.
+"""
+
+
+from luigi.target import FileAlreadyExists, FileSystem
+from luigi.contrib.hdfs.config import load_hadoop_cmd
+from luigi.contrib.hdfs import config as hdfs_config
+from luigi.contrib.hdfs import error as hdfs_error
+import logging
+import subprocess
+import datetime
+import os
+import re
+import warnings
+
+logger = logging.getLogger('luigi-interface')
+
+
+def create_hadoopcli_client():
+    """
+    Given that we want one of the hadoop cli clients (unlike snakebite),
+    this one will return the right one.
+    """
+    version = hdfs_config.get_configured_hadoop_version()
+    if version == "cdh4":
+        return HdfsClient()
+    elif version == "cdh3":
+        return HdfsClientCdh3()
+    elif version == "apache1":
+        return HdfsClientApache1()
+    else:
+        raise Exception("Error: Unknown version specified in Hadoop version"
+                        "configuration parameter")
+
+
+class HdfsClient(FileSystem):
+    """
+    This client uses Apache 2.x syntax for file system commands, which also matched CDH4.
+    """
+
+    recursive_listdir_cmd = ['-ls', '-R']
+
+    @staticmethod
+    def call_check(command):
+        p = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, close_fds=True, universal_newlines=True)
+        stdout, stderr = p.communicate()
+        if p.returncode != 0:
+            raise hdfs_error.HDFSCliError(command, p.returncode, stdout, stderr)
+        return stdout
+
+    def exists(self, path):
+        """
+        Use ``hadoop fs -stat`` to check file existence.
+        """
+
+        cmd = load_hadoop_cmd() + ['fs', '-stat', path]
+        logger.debug('Running file existence check: %s', u' '.join(cmd))
+        p = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, close_fds=True, universal_newlines=True)
+        stdout, stderr = p.communicate()
+        if p.returncode == 0:
+            return True
+        else:
+            not_found_pattern = "^.*No such file or directory$"
+            not_found_re = re.compile(not_found_pattern)
+            for line in stderr.split('\n'):
+                if not_found_re.match(line):
+                    return False
+            raise hdfs_error.HDFSCliError(cmd, p.returncode, stdout, stderr)
+
+    def rename(self, path, dest):
+        parent_dir = os.path.dirname(dest)
+        if parent_dir != '' and not self.exists(parent_dir):
+            self.mkdir(parent_dir)
+        if type(path) not in (list, tuple):
+            path = [path]
+        else:
+            warnings.warn("Renaming multiple files at once is not atomic.")
```

Comment:
It's probably not important for this use case, adding `stacklevel=2` would be nice.